fedimint-recurringd/src/db.rs

```diff
@@ -0,0 +1,185 @@
+            // TODO: clean up DB once parallel joins are enabled
+            error
+        }
+    })
+}
+
+pub async fn load_federation_client_databases(db: &Database) -> HashMap<FederationId, Database> {
+    load_federation_clients(db)
+        .await
+        .into_iter()
+        .map(|(federation_id, db_prefix)| (federation_id, open_client_db(db, db_prefix)))
+        .collect()
+}
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
+enum DbKeyPrefix {
+    ClientList = 0x00,
+    ClientDB = 0x01,
+    PaymentCodes = 0x02,
+    PaymentCodeNextInvoiceIndex = 0x03,
+    PaymentCodeInvoices = 0x04,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Encodable, Decodable)]
+pub struct FederationClientKey {
+    pub federation_id: FederationId,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Encodable, Decodable)]
+pub struct FederationClientPrefix;
+
+impl_db_record!(
+    key = FederationClientKey,
+    value = FederationClientEntry,
+    db_prefix = DbKeyPrefix::ClientList,
+);
+impl_db_lookup!(
+    key = FederationClientKey,
+    query_prefix = FederationClientPrefix
+);
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Encodable, Decodable)]
+pub struct FederationClientEntry {
+    pub db_prefix: FederationDbPrefix,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Encodable, Decodable)]
+pub struct PaymentCodeKey {
+    pub payment_code_id: PaymentCodeId,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Encodable, Decodable)]
+pub struct PaymentCodePrefix;
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Encodable, Decodable)]
+pub enum PaymentCodeVariant {
+    Lnurl { meta: String },
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Encodable, Decodable)]
+pub struct PaymentCodeEntry {
+    pub root_key: PaymentCodeRootKey,
+    pub federation_id: FederationId,
+    pub protocol: RecurringPaymentProtocol,
+    pub payment_code: String,
+    pub variant: PaymentCodeVariant,
+}
+
+impl_db_record!(
+    key = PaymentCodeKey,
+    value = PaymentCodeEntry,
+    db_prefix = DbKeyPrefix::PaymentCodes,
+);
+impl_db_lookup!(key = PaymentCodeKey, query_prefix = PaymentCodePrefix);
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Encodable, Decodable)]
+pub struct PaymentCodeNextInvoiceIndexKey {
+    pub payment_code_id: PaymentCodeId,
+}
+
+impl_db_record!(
+    key = PaymentCodeNextInvoiceIndexKey,
+    value = u64,
+    db_prefix = DbKeyPrefix::PaymentCodeNextInvoiceIndex,
+);
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Encodable, Decodable)]
+pub struct PaymentCodeInvoiceKey {
+    pub payment_code_id: PaymentCodeId,
+    pub index: u64,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Encodable, Decodable)]
+pub struct PaymentCodeInvoicePrefix {
+    payment_code_id: PublicKey,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Encodable, Decodable)]
+pub struct PaymentCodeInvoiceEntry {
+    pub operation_id: OperationId,
+    // TODO: use a more versatile type in the future to support BOLT12
+    pub invoice: Bolt11Invoice,
```

Comment:
Might want to make it an enum right now so its easier to migrate if BOLT12 is added